samples/oneapi/dpc/oneccl/sources/kmeans_distr_oneccl.cpp

```diff
@@ -0,0 +1,123 @@
+/*******************************************************************************
+ * Copyright 2021 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#include <CL/sycl.hpp>
+#include <iomanip>
+#include <iostream>
+#include <mpi.h>
+
+#ifndef ONEDAL_DATA_PARALLEL
+#define ONEDAL_DATA_PARALLEL
+#endif
+
+#include "oneapi/dal/algo/kmeans.hpp"
+#include "oneapi/dal/detail/oneccl/communicator.hpp"
+#include "oneapi/dal/detail/spmd_policy.hpp"
+#include "oneapi/dal/io/csv.hpp"
+
+#include "utils.hpp"
+
+namespace dal = oneapi::dal;
+
+template <typename Float>
+std::vector<dal::table>
+split_table_by_rows(const dal::detail::data_parallel_policy &p,
+                    const dal::table &t, std::int64_t split_count) {
+  ONEDAL_ASSERT(split_count > 0);
+  ONEDAL_ASSERT(split_count <= t.get_row_count());
+
+  const std::int64_t row_count = t.get_row_count();
+  const std::int64_t column_count = t.get_column_count();
+  const std::int64_t block_size_regular = row_count / split_count;
+  const std::int64_t block_size_tail = row_count % split_count;
+
+  std::vector<dal::table> result(split_count);
+
+  std::int64_t row_offset = 0;
+  for (std::int64_t i = 0; i < split_count; i++) {
+    const std::int64_t tail =
+        std::int64_t(i + 1 == split_count) * block_size_tail;
+    const std::int64_t block_size = block_size_regular + tail;
```

Comment:
it would be better to think out an approach for spreading tail across all workers, instead of assigning it to the last one

Reply:
Modified